File: algorithms/[102]二叉树的层次遍历/solution.py
# topics = ["树", "递归"]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from __future__ import annotations
from typing import Optional, List, Union


class TreeNode:
    def __init__(self, x: int = 0, left: Optional[TreeNode] = None, right: Optional[TreeNode] = None):
        self.val = x
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        result: List[List[int]] = []
        self.helper(result, root, 0)
        return result

    def helper(self, result: List[List[int]], node: Union[TreeNode, None], level: int) -> None:
        if not node:
            return
        if len(result) <= level:
            result.append([])
        result[level].append(node.val)
        self.helper(result, node.left, level + 1)
        self.helper(result, node.right, level + 1)

File: algorithms/[102]二叉树的层次遍历/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
const levelOrder = function(root) {
    const result = [];
    const helper = function(node, level) {
        if (!node) return;
        if (result.length <= level) result.push([]);
        result[level].push(node.val);
        helper(node.left, level + 1);
        helper(node.right, level + 1);
    }
    helper(root, 0);
    return result;
};
